### PR TITLE
Add COUNT_DISTINCT aggregation

### PR DESCRIPTION
**Pull request by esadler-hbo**
## Description

Add COUNT_DISTINCT aggregation in Feathr window aggregations.

## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure to clarify your proposed changes.

### CODE REVIEW

File: src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala [hunk 1]
```diff
@@ -983,4 +983,87 @@ class SlidingWindowAggIntegTest extends FeathrIntegTest {
...
+
+
+    /**
+     * Expected output:
+     * +--------+----+----+
+     * |x|   f|   g|
+     * +--------+----+----+
+     * |       1|   6|   2|
+     * |       2|   6|   2|
+     * |       3|   1|   1|
```
Comment (hangfei) [inline src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala:1050]: doesn't match the code(expectedRows)

**Review by hangfei — COMMENTED** [review]
thanks for the PR!

File: src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala (hunk 1 above)
Comment (esadler-hbo) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: @hangfei good catch. I loaded the test data into python to double check the results are expected. Curious how your team validates the results.

Comment (hangfei) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: Are you asking about unit test?
@ahlag added a pretty nice dev guide for scala unit test: https://github.com/linkedin/feathr/blob/main/docs/dev_guide/scala_dev_guide.md. You can try it out.

We also have github actions to kick off e2e testing.

Comment (esadler-hbo) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: Sweet. I will take a look!

Comment (ahlag) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: @esadler-hbo 
Please let me know if there’s any question or anything I left out.

File: src/main/scala/com/linkedin/feathr/swj/aggregate/CountDistinctAggregate.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package com.linkedin.feathr.swj.aggregate
+
+import com.linkedin.feathr.swj.aggregate.AggregationType._
+import org.apache.spark.sql.types._
+
+/**
+  * MAX aggregation implementation.
```
Comment (xiaoyongzhu) [inline src/main/scala/com/linkedin/feathr/swj/aggregate/CountDistinctAggregate.scala:7]: comment is wrong

File: src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala [hunk 3]
```diff
@@ -983,4 +983,87 @@ class SlidingWindowAggIntegTest extends FeathrIntegTest {
...
+        StructField(features.head, LongType), // f
+        StructField(features.last, LongType) // g
+      ))
+
+    val expectedRows = Array(
+      new GenericRowWithSchema(Array(1, 6, 2), expectedSchema),
+      new GenericRowWithSchema(Array(2, 5, 2), expectedSchema),
+      new GenericRowWithSchema(Array(3, 1, 1), expectedSchema))
+    val dfs = runLocalFeatureJoinForTest(featureJoinAsString, featureDefAsString, "featuresWithFilterObs.avro.json").data
+    dfs.show()
```
Comment (jaymo001) [inline src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala:1065]: nit: delete it.

File: src/main/scala/com/linkedin/feathr/swj/aggregate/CountDistinctAggregate.scala (hunk 2 above)
Comment (esadler-hbo) [reply src/main/scala/com/linkedin/feathr/swj/aggregate/CountDistinctAggregate.scala]: Fixed

File: src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala (hunk 3 above)
Comment (esadler-hbo) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: Deleted

**Review by xiaoyongzhu — APPROVED** [review]

File: src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala (hunk 1 above)
Comment (esadler-hbo) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: @ahlag super helpful doc! Thank you!

**Review by jaymo001 — APPROVED** [review]

File: src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala (hunk 1 above)
Comment (xiaoyongzhu) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: @esadler-hbo fixed in the below commit (https://github.com/linkedin/feathr/pull/594/commits/1db050c7b9692158bd9ecdcf87ae9e41763900e0)

Comment (xiaoyongzhu) [reply src/test/scala/com/linkedin/feathr/offline/SlidingWindowAggIntegTest.scala]: resolving the conversation now

**Pull request merged**